### PR TITLE
Fix unused parameters in iocore

### PR DESCRIPTION
**Pull request by freak82**
Use `/* name ATS_UNUSED */` for fully unused parameters.
Use `[[maybe_unused]]` for conditionally unused parameters.

This pull request is part of the effort for [removing](https://github.com/apache/trafficserver/issues/11377) the `-Wno-unused-parameter` warning suppression.

### CODE REVIEW

**Comment by JosiahWI** [comment]
Looks like the `len` parameter in `RamCacheLRU::put` is used in a debug call.

**Comment by freak82** [comment]
@JosiahWI 
You are right. Thank you.
 I forgot to build the project in `Debug` mode. Now it should be OK.

**Comment by JosiahWI** [comment]
[approve ci rocky]

**Review by cmcfarlen — APPROVED** [review]

File: src/iocore/net/TLSSessionResumptionSupport.cc [hunk 1]
```diff
@@ -191,7 +191,7 @@ TLSSessionResumptionSupport::getSession(SSL *ssl, const unsigned char *id, int l
 }
 
 std::shared_ptr<SSL_SESSION>
-TLSSessionResumptionSupport::getOriginSession(SSL *ssl, const std::string &lookup_key)
+TLSSessionResumptionSupport::getOriginSession(SSL * /* ssl ATS_UNUSED */, const std::string &lookup_key)
```
Comment (JosiahWI) [inline src/iocore/net/TLSSessionResumptionSupport.cc:194]: Can we remove the `ssl` parameter?

File: src/iocore/net/TLSSessionResumptionSupport.cc [hunk 2]
```diff
@@ -221,12 +221,14 @@ TLSSessionResumptionSupport::clear()
 
 #ifdef HAVE_SSL_CTX_SET_TLSEXT_TICKET_KEY_EVP_CB
 int
-TLSSessionResumptionSupport::_setSessionInformation(ssl_ticket_key_block *keyblock, SSL *ssl, unsigned char *keyname,
-                                                    unsigned char *iv, EVP_CIPHER_CTX *cipher_ctx, EVP_MAC_CTX *hctx)
+TLSSessionResumptionSupport::_setSessionInformation(ssl_ticket_key_block *keyblock, SSL * /* ssl ATS_UNUSED */,
```
Comment (JosiahWI) [inline src/iocore/net/TLSSessionResumptionSupport.cc:224]: Here too.

File: src/iocore/net/TLSSessionResumptionSupport.cc [hunk 3]
```diff
@@ -221,12 +221,14 @@ TLSSessionResumptionSupport::clear()
...
+                                                    unsigned char *keyname, unsigned char *iv, EVP_CIPHER_CTX *cipher_ctx,
+                                                    EVP_MAC_CTX *hctx)
 #else
 int
-TLSSessionResumptionSupport::_setSessionInformation(ssl_ticket_key_block *keyblock, SSL *ssl, unsigned char *keyname,
-                                                    unsigned char *iv, EVP_CIPHER_CTX *cipher_ctx, HMAC_CTX *hctx)
+TLSSessionResumptionSupport::_setSessionInformation(ssl_ticket_key_block *keyblock, SSL * /* ssl ATS_UNUSED */,
```
Comment (JosiahWI) [inline src/iocore/net/TLSSessionResumptionSupport.cc:229]: and here.

File: src/iocore/net/SSLUtils.cc [hunk 4]
```diff
@@ -946,7 +946,8 @@ SSLMultiCertConfigLoader::default_server_ssl_ctx()
 }
 
 static bool
-SSLPrivateKeyHandler(SSL_CTX *ctx, const SSLConfigParams *params, const char *keyPath, const char *secret_data, int secret_data_len)
+SSLPrivateKeyHandler(SSL_CTX *ctx, const SSLConfigParams * /* params ATS_UNUSED */, const char *keyPath, const char *secret_data,
```
Comment (JosiahWI) [inline src/iocore/net/SSLUtils.cc:949]: Can we remove the `params` parameter entirely?

File: src/iocore/net/ALPNSupport.cc [hunk 5]
```diff
@@ -88,7 +88,7 @@ ALPNSupport::setSelectedProtocol(const unsigned char *proto, unsigned int len)
 }
 
 int
-ALPNSupport::advertise_next_protocol(SSL *ssl, const unsigned char **out, unsigned *outlen)
+ALPNSupport::advertise_next_protocol(SSL * /* ssl ATS_UNUSED */, const unsigned char **out, unsigned *outlen)
```
Comment (JosiahWI) [inline src/iocore/net/ALPNSupport.cc:91]: Can the `ssl` parameter be removed?

File: src/iocore/net/ALPNSupport.cc [hunk 6]
```diff
@@ -98,8 +98,8 @@ ALPNSupport::advertise_next_protocol(SSL *ssl, const unsigned char **out, unsign
 }
 
 int
-ALPNSupport::select_next_protocol(SSL *ssl, const unsigned char **out, unsigned char *outlen, const unsigned char *in,
-                                  unsigned inlen)
+ALPNSupport::select_next_protocol(SSL * /* ssl ATS_UNUSED */, const unsigned char **out, unsigned char *outlen,
```
Comment (JosiahWI) [inline src/iocore/net/ALPNSupport.cc:101]: Can the `ssl` parameter be removed?

File: src/iocore/eventsystem/MIOBufferWriter.cc [hunk 7]
```diff
@@ -76,7 +76,7 @@ MIOBufferWriter::operator>>(std::ostream &stream) const
 }
 
 ssize_t
-MIOBufferWriter::operator>>(int fd) const
+MIOBufferWriter::operator>>(int /* fd ATS_UNUSED */) const
 {
   return 0;
 }
```
Comment (JosiahWI) [inline src/iocore/eventsystem/MIOBufferWriter.cc:82]: Um... interesting. Can this whole method be removed I wonder?

File: include/iocore/eventsystem/Lock.h [hunk 8]
```diff
@@ -240,7 +240,7 @@ class ProxyMutex : public RefCountObj
 
   */
   void
-  init(const char *name = "UnnamedMutex")
+  init(const char * /* name  ATS_UNUSED */ = "UnnamedMutex")
```
Comment (JosiahWI) [inline include/iocore/eventsystem/Lock.h:243]: Can the `name` parameter be removed?

**Review by JosiahWI — CHANGES_REQUESTED** [review]
Thanks for the iocore cleanup. In a few of these cases it looks like maybe the parameter could be removed.

**Comment by freak82** [comment]
@JosiahWI 
I can remove them by the end of  this week. Should I wait for another approval to remove them or I can just do it?

**Comment by JosiahWI** [comment]
Since the goal of this PR is to move towards removing the `-Wno-unused-parameter` suppression, we would be fine with merging this as-is. If you would like to go ahead and remove the parameters, you can do so right away, in this PR or in a separate one. If you'd rather not, I have time to remove them myself in a separate PR. What would you prefer?

File: src/iocore/eventsystem/MIOBufferWriter.cc (hunk 7 above)
Comment (freak82) [reply src/iocore/eventsystem/MIOBufferWriter.cc:82]: I'm not sure about this one.
It's inside
```
#if defined(UNIT_TEST_BUFFER_WRITER)
...
#else
```
and in the `#else` branch this same method has an actual implementation.

Comment (JosiahWI) [reply src/iocore/eventsystem/MIOBufferWriter.cc:82]: Sure enough, it's for the unit test. Let's leave this one as you have it for now.

**Comment by freak82** [comment]
One additional question.
The functions which accept `SSL` parameter are called, at least few that I checked, by other functions which also accept such parameter, etc.
Should I remove this parameter from all levels of the call-stack? The changes may spread to lots of files depending on the call-stacks.

**Comment by JosiahWI** [comment]
Thanks for asking about this. It should be fine. Some of them are SSL callbacks that require the parameter, so you probably won't need to remove them in very many places, but they might need to be marked with `ATS_UNUSED` in some of these callbacks.

**Comment by freak82** [comment]
@JosiahWI 
The requested parameters were completely removed.
Hopefully I didn't miss something.

**Comment by JosiahWI** [comment]
The AuTest slice_prefetch failed.
```
        Reason: File differences
         Gold File : /home/jenkins/workspace/Github_Builds/autest/src/tests/gold_tests/pluginTest/slice/gold/slice_prefetch.gold
         Data File : /tmp/sandbox/slice_prefetch/ts/log/cache.log
           bytes 0-6/18 miss
           bytes 7-13/18 miss
           bytes 14-17/18 miss
           bytes 14-17/18 hit-fresh
           - miss, none
           bytes 0-6/18 hit-fresh
           bytes 7-13/18 hit-fresh
           bytes 14-17/18 hit-fresh
           - hit-fresh, none
         - bytes 0-6/18 hit-stale
         - bytes ``/18 hit-stale
         - bytes ``/18 hit-stale
         - bytes 14-17/18 hit-fresh
         - - hit-stale, none
         - bytes 0-6/18 hit-fresh
         - bytes 7-13/18 hit-fresh
         - bytes 14-17/18 hit-fresh
         - bytes 0-17/18 hit-fresh, none
         - bytes 0-4/18 miss
         - bytes ``/18 miss
         - bytes ``/18 miss
         - bytes ``/18 miss
         - bytes 15-17/18 hit-fresh
         - bytes 5-16/18 miss, none
         - bytes 0-6/18 hit-fresh
         - */18 hit-fresh, none
    ```

**Comment by JosiahWI** [comment]
[approve ci autest]

**Review by JosiahWI — APPROVED** [review]
Thank you for the additional cleanup! This looks great!

**Review by cmcfarlen — APPROVED** [review]

**Comment by cmcfarlen** [comment]
Cherry-picked to v10.0.x

**Pull request merged**